File: 628-maximum-product-of-three-numbers/maximum-product-of-three-numbers.py
class Solution:
    def maximumProduct(self, nums: List[int]) -> int:
        n=len(nums)

        mx1=mx2=mx3=float('-inf')
        mn1=mn2=float('inf');
        for num in nums:
            if num>=mx1:
                mx3=mx2
                mx2=mx1
                mx1=num
            elif num>=mx2:
                mx3=mx2
                mx2=num
            elif num>=mx3:
                mx3=num

            if num<=mn1:
                mn2=mn1
                mn1=num
            elif num<=mn2:
                mn2=num
    
        prod1=mx1*mx2*mx3
        prod2=mn1*mn2*mx1

        return max(prod1, prod2)

File: 628-maximum-product-of-three-numbers/maximum-product-of-three-numbers.java
class Solution{
    public int maximumProduct(int[] nums){
        int n=nums.length;

        int mx1=Integer.MIN_VALUE, mx2=Integer.MIN_VALUE, mx3=Integer.MIN_VALUE;
        int mn1=Integer.MAX_VALUE, mn2=Integer.MAX_VALUE;
        for(int num: nums){
            if(num>=mx1){
                mx3=mx2;
                mx2=mx1;
                mx1=num;
            }
            else if(num>=mx2){
                mx3=mx2;
                mx2=num;
            }
            else if(num>=mx3){
                mx3=num;
            }

            if(num<=mn1){
                mn2=mn1;
                mn1=num;
            }
            else if(num<=mn2){
                mn2=num;
            }
        }

        int prod1=mx1*mx2*mx3;
        int prod2=mn1*mn2*mx1;

        return Math.max(prod1, prod2);
    }
}